SonarQube.Client/Requests/UnversionedRequestFactory.cs

```diff
@@ -0,0 +1,84 @@
+﻿/*
+ * SonarQube Client
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using SonarQube.Client.Logging;
+
+namespace SonarQube.Client.Requests
+{
+    internal class UnversionedRequestFactory : IRequestFactory
+    {
+        private readonly ILogger logger;
+
+        private readonly Dictionary<Type, Func<IRequest>> requestToFactoryMap;
+
+        public UnversionedRequestFactory(ILogger logger)
+        {
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            requestToFactoryMap = new Dictionary<Type, Func<IRequest>>();
+            this.logger = logger;
+        }
+
+        /// <summary>
+        /// Registers a simple request factory
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request interface to create implementation for.</typeparam>
+        /// <typeparam name="TRequestImpl">The type implementing TRequest that has a parameterless constructor.</typeparam>
+        public UnversionedRequestFactory RegisterRequest<TRequest, TRequestImpl>()
+            where TRequest : IRequest
+            where TRequestImpl : TRequest, new()
+        {
+            if (requestToFactoryMap.ContainsKey(typeof(TRequest)))
+            {
+                logger.Error($"Registration for {typeof(TRequest).Name} already exists.");
+                throw new InvalidOperationException(
+                    $"Registration for {typeof(TRequest).Name} already exists.");
+            }
+
+            requestToFactoryMap[typeof(TRequest)] = () => new TRequestImpl();
+            logger.Debug($"Registered {typeof(TRequestImpl).FullName}");
+
+            return this;
+        }
+
+        public TRequest Create<TRequest>(ServerInfo serverInfo) where TRequest : IRequest
```

Comment:
I think the ServerInfo class can go now: it's not needed at all for SC, and for SQ we can pass the version in the ctor - the version would be the same for every `Create` call, as we would create a new factory each time the connection changes.

Reply:
As discussed offline, this would require changing the factory to make calls to the server. I'll add some tickets to the backlog.